src/test/reporters/github.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-ignore
+import milliseconds from 'ms';
+import path from 'path';
+import { BaseReporter, formatFailure, workspaceRelativePath } from './base';
+import {
+  TestCase,
+  FullResult,
+} from '../../../types/testReporter';
+
+type GithubLogType = 'debug' | 'notice' | 'warning' | 'error';
+
+type GithubLogOptions = Partial<{
+  title: string;
+  file: string;
+  col: number;
+  endColumn: number;
+  line: number;
+  endLine: number;
+}>;
+
+class GithubLogger {
+	private _isCI: boolean = process.env.CI === 'true';
+  private _isGithubAction: boolean = process.env.GITHUB_ACTION !== undefined;
+  private _shouldLog = (this._isCI && this._isGithubAction) || process.env.PW_GH_ACTION_DEBUG === 'true' ;
+
+  private _log(
+    message: string,
+    type: GithubLogType = 'notice',
+    options: GithubLogOptions = {}
+  ) {
+    if (this._shouldLog) {
+      if (this._isGithubAction) message = message.replace(/\n/g, '%0A');
+
+      const configs = Object.entries(options)
+          .map(([key, option]) => `${key}=${option}`)
+          .join(',');
+      console.log(`::${type} ${configs}::${message}`);
+    }
+  }
+
+  debug(message: string, options?: GithubLogOptions) {
+    this._log(message, 'debug', options);
+  }
+
+  error(message: string, options?: GithubLogOptions) {
+    this._log(message, 'error', options);
+  }
+
+  notice(message: string, options?: GithubLogOptions) {
+    this._log(message, 'notice', options);
+  }
+
+  warning(message: string, options?: GithubLogOptions) {
+    this._log(message, 'warning', options);
+  }
+}
+
+export class GithubReporter extends BaseReporter {
+  githubLogger = new GithubLogger();
+
+  override async onEnd(result: FullResult) {
+    super.onEnd(result);
+    this.epilogue(true);
+  }
+
+  protected override printSlowTests() {
+    this.getSlowTests().forEach(([file, duration]) => {
+      const filePath = workspaceRelativePath(
+          path.join(process.cwd(), file)
+      );
+      this.githubLogger.warning(`${filePath} (${milliseconds(duration)})`, {
+        title: 'Slow Test',
+        file: filePath,
+      });
+    });
+
+
+  }
+
+  protected override printSummary(summary: string){
```

Comment:
It would be considered a better design if instead of overriding here and below, you would _implement_ reporter interface via defining the methods like `onEnd` and calling helper methods like `epilogue` from `base`. You are half way there.

Reply:
I didn't quite get the idea behind that. If I'm not inheriting `base`, I'd have to implement all the methods like `onEnd`, `onBegin`, etc, keep a copy of base, call it's functions and then call epilogue at the end. 
But even then, I can't modify the behaviour of epilogue (which is to print to console).

What I intended here is to follow the same principal as `line`, `dot`, and `list` reporter, keep all the logic in base and simply override the function which prints to console as the output format is the only thing different here for GH action.  

Can you please give an example of how I could do the suggested method?

Reply:
Sure! If you look at `line`, `dot`, `list`, you'll see that they only override interface methods from `base`, they don't override protected methods in `base`:

```js
// I'm in dot.ts, I'm handling onEnd
override async onEnd(result: FullResult) {
  // First, I'll delegate it to base so that it could process this signal and update its state.
  await super.onEnd(result);
  // Then I'll do dot-specific logic
  process.stdout.write('\n');
  // And also use some utilities from base that would do something that
  // is reused in many reporters.
  this.epilogue(true);
}
```

To reiterate:
1. Handle signal
2. Let base know to update the state
3. Use utilities in base to produce desired outcome

With this design, all control flows are direct github -> base. It allows quick switching from inheritance to aggregation. With your proposed design, the calls are github -> base -> github (via override delegation).

Reply:
I was trying to avoid duplicating the epilogue logic.
Removed the overloads now.
Thanks !